test/test_event_detectors.py

```
@@ -1,133 +1,346 @@
             translation_detector.update_with_latest_motion_data()
 
-            # wait one timestep to detect that it is moving
-            time.sleep(translation_detector.get_n_changes_wait_time(1))
+            print(f"Step {i+1}/{steps} - Milk position: {interp_pos}")
+            print(
+                f"  Latest TranslationEvent: {milk_tracker.get_latest_event_of_type(TranslationEvent)}"
+            )
+            print(
+                f"  Latest StopTranslationEvent: {milk_tracker.get_latest_event_of_type(StopTranslationEvent)}"
+            )
 
-            translation_event = milk_tracker.get_latest_event_of_type(TranslationEvent)
-            self.assertTrue(translation_event is not None)
+            time.sleep(translation_detector.get_n_changes_wait_time(1) * 2)
 
-            # update once to detect two consistent gradients of zero value between last three updates.
+        # Simulate no movement to trigger StopTranslationEvent
+        for j in range(5):
             translation_detector.update_with_latest_motion_data()
+            world.notify_state_change()
+            print(
+                f"Wait {j+1}/5 - Latest StopTranslationEvent: {milk_tracker.get_latest_event_of_type(StopTranslationEvent)}"
+            )
+            time.sleep(translation_detector.get_n_changes_wait_time(1) * 2)
 
-            # wait one timestep to detect that it is not moving
-            time.sleep(translation_detector.get_n_changes_wait_time(1))
+        assert milk_tracker.get_latest_event_of_type(TranslationEvent) is not None
+        assert milk_tracker.get_latest_event_of_type(StopTranslationEvent) is not None
 
-            self.assertTrue(milk_tracker.get_first_event_of_type_after_event(StopTranslationEvent, translation_event)
-                            is not None)
-        except Exception as e:
-            raise e
-        finally:
+    finally:
+        # Always stop and join the detector thread
+        try:
```

Don't try except Exception like that, this ignores the errors that happen when they should not be ignored.